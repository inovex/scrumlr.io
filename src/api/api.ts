import Parse from "parse";

const callApi = <RequestType, ResponseType>(endpoint: string, request: RequestType) => {
    return Parse.Cloud.run(endpoint, request) as Promise<ResponseType>;
}

export const Api = {
    createBoard: () => {
        return callApi<{}, string>('createBoard', {});
    },
    joinBoard: (boardId: string) => {
        return callApi<{ boardId : string }, { status: 'accepted' | 'rejected' | 'pending', joinRequestReference?: string }>('joinBoard', { boardId});
    },
    addNote: (boardId: string, text: string) => {
        return callApi<{ boardId: string, text: string }, boolean>('addNote', { boardId, text });
    },
    deleteNote: (boardId: string, noteId: string) => {
        return callApi<{ boardId: string, noteId: string}, boolean>('deleteNote', { boardId, noteId });
    },
    editNote: (boardId: string, noteId: string, text: string) => {
        return callApi<{ boardId: string, noteId: string, text: string}, boolean>('editNote', { boardId, noteId, text });
    }
}