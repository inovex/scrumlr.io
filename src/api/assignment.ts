import {SERVER_HTTP_URL} from "../config";

export const AssignmentAPI = {
  createAssignment: async (boardId: string, noteId: string, name: string) => {
    try {
      const response = await fetch(`${SERVER_HTTP_URL}/boards/${boardId}/assignments`, {
        method: "POST",
        credentials: "include",
        body: JSON.stringify({
          note: noteId,
          name,
        }),
      });

      if (response.status === 201) {
        return;
      }

      throw new Error(`add assignee request resulted in status ${response.status}`);
    } catch (error) {
      throw new Error(`unable to create board: ${error}`);
    }
  },
  deleteAssignment: async (boardId: string, assignmentId: string) => {
    try {
      await fetch(`${SERVER_HTTP_URL}/boards/${boardId}/assignments/${assignmentId}`, {
        method: "DELETE",
        credentials: "include",
      });
    } catch (error) {
      throw new Error(`unable to delete assignment: ${error}`);
    }
  },
};
