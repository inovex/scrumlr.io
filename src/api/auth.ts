import {callAPI} from "./index";

const generateState = (prefix: string) => {
  // generate random state id and store origin into the session storage
  const state = `${prefix}-${Math.random().toString(36).substring(2, 15) + Math.random().toString(36).substring(2, 15)}`;
  sessionStorage.setItem(state, window.location.href);
  return state;
};

export const AuthAPI = {
  /**
   * Sign in with Google.
   *
   * @returns the redirection URL
   */
  signInWithGoogle: () => callAPI<{state: string}, string>("GoogleSignIn", {state: generateState("google")}),

  signInWithGithub: () => callAPI<{state: string}, string>("GithubSignIn", {state: generateState("github")}),

  /**
   * Verify the sign in with Google by the specified code.
   *
   * @param code the verification code returned by the authentication provider
   * @param state the state passed by the authentication provider (see https://auth0.com/docs/protocols/state-parameters)
   *
   * @returns the user information
   */
  verifyGoogleSignIn: async (code: string, state: string) => {
    // check if state is available in storage and execute call on match
    const redirectURL = sessionStorage.getItem(state);
    if (redirectURL) {
      const user = await callAPI<{code: string}, {id: string; name: string; accessToken: string; idToken: string}>("GoogleVerifySignIn", {code});
      return {
        user,
        redirectURL,
      };
    }
    throw new Error("state does not match");
  },

  /**
   * Verify the sign in with Google by the specified code.
   *
   * @param code the verification code returned by the authentication provider
   * @param state the state passed by the authentication provider (see https://auth0.com/docs/protocols/state-parameters)
   *
   * @returns the user information
   */
  verifyGithubSignIn: async (code: string, state: string) => {
    // check if state is available in storage and execute call on match
    const redirectURL = sessionStorage.getItem(state);
    if (redirectURL) {
      const user = await callAPI<{code: string}, {id: string; name: string; accessToken: string}>("GithubVerifySignIn", {code});
      return {
        user,
        redirectURL,
      };
    }
    throw new Error("state does not match");
  },
};
