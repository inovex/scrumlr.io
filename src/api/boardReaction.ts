import {SERVER_HTTP_URL} from "config";
import {ReactionType} from "types/reaction";
import {BoardReactionType} from "types/boardReaction";

export const BoardReactionAPI = {
  addBoardReaction: async (boardId: string, reactionType: ReactionType) => {
    try {
      const response = await fetch(`${SERVER_HTTP_URL}/boards/${boardId}/board-reactions`, {
        method: "POST",
        credentials: "include",
        body: JSON.stringify({
          reactionType,
        }),
      });

      if (response.status === 201) {
        return (await response.json()) as BoardReactionType;
      }

      throw new Error(`create board reaction request resulted in status ${response.status}`);
    } catch (error) {
      throw new Error(`unable to create board reaction with error: ${error}`);
    }
  },
};
