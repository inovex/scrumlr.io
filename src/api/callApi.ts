import Parse from "parse";

/**
 * Helper function for calls on the cloud functions of the Parse server backend.
 *
 * @param endpoint the name of the endpoint
 * @param request the request parameters
 *
 * @returns the asynchronous server response wrapped in a `Promise`
 */
export const callAPI = <RequestType, ResponseType>(endpoint: string, request: RequestType) => Parse.Cloud.run(endpoint, request) as Promise<ResponseType>;
