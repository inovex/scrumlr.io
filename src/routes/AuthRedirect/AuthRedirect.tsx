import queryString from "query-string";
import {useEffect, useState} from "react";
import Parse from "parse";
import {API} from "api";
import LoadingScreen from "components/LoadingScreen/LoadingScreen";
import {useLocation} from "react-router";
import ErrorPage from "components/ErrorPage/ErrorPage";

function AuthRedirect() {
  const [status, setStatus] = useState<{error?: string; redirect?: string}>({});
  const location = useLocation();
  const params = queryString.parse(location.search);

  useEffect(() => {
    if (params.error) {
      setStatus({error: params.error as string});
    } else if (params.code && params.state) {
      if ((params.state as string).startsWith("google")) {
        API.verifyGoogleSignIn(params.code as string, params.state as string)
          .then((res) => {
            const user = new Parse.User();
            const authData = {
              id: res.user.id,
              id_token: res.user.idToken,
              access_token: res.user.accessToken,
            };

            user.linkWith("google", {authData}).then(() => {
              user.set("displayName", res.user.name);
              user.save().then(() => {
                window.location.href = res.redirectURL;
              });
            });
          })
          .catch(() => {
            setStatus({error: "State does not match"});
          });
      }

      if ((params.state as string).startsWith("github")) {
        API.verifyGithubSignIn(params.code as string, params.state as string)
          .then((res) => {
            const user = new Parse.User();
            const authData = {
              id: res.user.id,
              access_token: res.user.accessToken,
            };

            user.linkWith("github", {authData}).then(() => {
              user.set("displayName", res.user.name);
              user.save().then(() => {
                window.location.href = res.redirectURL;
              });
            });
          })
          .catch(() => {
            setStatus({error: "State does not match"});
          });
      }
    } else {
      setStatus({error: "Not a valid entrypoint"});
    }
  }, [params.code, params.error, params.state]);

  if (status.error) {
    return <ErrorPage errorMessage="Oops! Unable to sign in." />;
  }

  return <LoadingScreen />;
}

export default AuthRedirect;
