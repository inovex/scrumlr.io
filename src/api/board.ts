import {Color} from "constants/colors";
import {callAPI} from "./callApi";

export const BoardAPI = {
  /**
   * Creates a board with the specified parameters and returns the board id.
   *
   * @param columns the definition of the columns
   * @param name the board name
   *
   * @returns the board id of the created board
   */
  createBoard: (columns: {name: string; hidden: boolean; color: Color}[], name?: string) =>
    callAPI<{columns: {name: string; hidden: boolean}[]; name?: string}, string>("createBoard", {columns, name}),

  /**
   * Create join request for a board session.
   * The return value might have the status `accepted` (user is permitted to join the board), `rejected` (the join
   * request of the user was rejected by an admin) or `pending`. If the state is pending the response will include
   * the reference on the join request state in the attribute `joinRequestReference`.
   *
   * @param boardId the board id
   *
   * @returns `true` if the operation succeeded or throws an error otherwise
   */
  joinBoard: (boardId: string) => callAPI<{boardId: string}, {status: "accepted" | "rejected" | "pending"; joinRequestReference?: string}>("joinBoard", {boardId}),
  acceptJoinRequests: (boardId: string, userIds: string[]) => callAPI<{board: string; users: string[]}, boolean>("acceptUsers", {board: boardId, users: userIds}),
  rejectJoinRequests: (boardId: string, userIds: string[]) => callAPI<{board: string; users: string[]}, boolean>("rejectUsers", {board: boardId, users: userIds}),
};
