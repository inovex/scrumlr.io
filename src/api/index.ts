import {BoardAPI} from "./board";
import {NoteAPI} from "./note";
import {ColumnAPI} from "./column";

/** This class lists all API functions of the server. */
export const API = {
  ...BoardAPI,
  ...ColumnAPI,
  ...NoteAPI,
};
