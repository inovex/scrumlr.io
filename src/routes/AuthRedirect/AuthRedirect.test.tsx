import {API} from "api";
import {render} from "@testing-library/react";
import {createMemoryHistory} from "history";
import {Router} from "react-router";
import AuthRedirect from "./AuthRedirect";

jest.mock("api", () => ({
    API: {
      verifyGoogleSignIn: jest.fn(),
    },
  }));

beforeEach(() => {
  (API.verifyGoogleSignIn as jest.Mock).mockClear();
});

describe("AuthRedirect", () => {
  test("should render loading screen while waiting for server", () => {
    (API.verifyGoogleSignIn as jest.Mock).mockReturnValue({then: () => ({catch: jest.fn()})});

    const history = createMemoryHistory({initialEntries: ["/test?code=1&state=1"], initialIndex: 0});
    const result = render(
      <Router history={history}>
        <AuthRedirect />
      </Router>
    ).container;
    expect(API.verifyGoogleSignIn).toHaveBeenCalled();
    expect(result.querySelector(".loading-screen")).toBeInTheDocument();
  });

  test("should show error page if redirect result contains an error", () => {
    const history = createMemoryHistory({initialEntries: ["/error=1"], initialIndex: 0});

    const result = render(
      <Router history={history}>
        <AuthRedirect />
      </Router>
    ).container;

    expect(result.querySelector(".error-page")).toBeInTheDocument();
  });

  test("should show error page if page was called with invalid params", () => {
    const history = createMemoryHistory();

    const result = render(
      <Router history={history}>
        <AuthRedirect />
      </Router>
    ).container;

    expect(result.querySelector(".error-page")).toBeInTheDocument();
  });
});
