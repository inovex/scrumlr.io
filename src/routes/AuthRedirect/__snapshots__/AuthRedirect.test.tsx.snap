// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthRedirect should show error page if page was called with invalid params 1`] = `
<div>
  <div
    class="error-page"
  >
    <span>
      Oops! Unable to sign in.
    </span>
  </div>
</div>
`;

exports[`AuthRedirect should show error page if redirect result contains an error 1`] = `
<div>
  <div
    class="error-page"
  >
    <span>
      Oops! Unable to sign in.
    </span>
  </div>
</div>
`;
